### PR TITLE
feat: add thread message posting

### PR DESCRIPTION
**Pull request by atanass**
Description:
Add a new parameter to the `notify` command: `thread` to enable Slack message replies in a thread.

Motivation:
https://github.com/CircleCI-Public/slack-orb/issues/378

Checklist:
[x] All new jobs, commands, executors, parameters have descriptions.
[ ] Usage Example version numbers have been updated.
[ ] Changelog has been updated.

### CODE REVIEW

**Comment by atanass** [comment]
I've done some testing and it seems like it's working relatively well. Evidence below:
Pipeline structure with 2 message threads (1 for building & 1 for deploying)
image

Messages in Slack:
image
image
image

**Comment by atanass** [comment]
Test with 3 consequent messages in a thread.
Snippet:
```
version: 2.1

orbs:
  slack: nukengprod/ecr@dev:slack

executors:
  default:
    docker:
      - image: cimg/base:2023.10

jobs:
  test:
    executor: default
    steps:
      - run: echo testing
  build:
    executor: default
    steps:
      - run: echo building
  deploy:
    executor: default
    steps:
      - run: echo deploying

  slack:
    executor: default
    parameters:
      finished:
        type: boolean
        default: false
      phase:
        type: string
    steps:
      - run:
          name: Prepare slack template
          command: |
            echo "{\"blocks\": [{\"type\": \"section\",\"text\": {\"type\": \"plain_text\",\"text\": \"Deployment process: << parameters.phase >>.\",\"emoji\": true}}]}" > template.json
            echo "export SLACK_PIPELINE_DEPLOY_TEMPLATE=\"\$(cat template.json)\"" >> $BASH_ENV
      - slack/notify:
          channel: "#atanas-test"
          template: SLACK_PIPELINE_DEPLOY_TEMPLATE
          event: always
          thread: build

workflows:
  version: 2
  test:
    jobs:
      - slack:
          name: pre-build
          phase: start
      - build:
          requires:
            - pre-build
      - slack:
          name: build-in-progress
          phase: "in progress"
          requires:
            - pre-build
      - slack:
          name: post-build
          phase: finished
          requires:
            - build
```

Pipeline:
image


Messages:
image

**Review by aaronstillwell — COMMENTED** [review]
Thanks for your work here @atanass. I've added a few comments for your review. Tomorrow I'll see about verifying some of your tests on my end and will come circle back!

File: src/commands/notify.yml [hunk 1]
```diff
@@ -77,6 +77,11 @@ parameters:
     type: string
     default: Slack - Sending Notification
     description: Specify a custom step name for this command, if desired
+  thread:
+    type: string
+    default: ""
+    description: |
+      Name of the thread in the pipeline. Enables threads.
```
[the remark this answers is not in the document]
Comment (atanass) [reply src/commands/notify.yml]: Thank you very much for the perspective. I've updated the description as per your proposal and the name of the parameter to `thread_id`

File: src/commands/notify.yml [hunk 2]
```diff
@@ -111,3 +124,12 @@ steps:
...
       command: <<include(scripts/main.sh)>>
+  - when:
+      condition:
+        not:
+          equal: [ "", <<parameters.thread>> ]
+      steps:
+        - persist_to_workspace:
+            root: /tmp/SLACK_THREAD_INFO
+            paths:
+              - .
```
[the remark this answers is not in the document]
Comment (atanass) [reply src/commands/notify.yml:137]: Yes. If a user decides to mark different notifications with the same thread id, but they're not really interdependent, they might end up with a confusing thread of messages. In such cases users should omit the `thread_id` parameter, whose absence makes the orb command fall back to its default behaviour of posting single messages. I think properly describing the intent of the parameter could prevent from such corner-case implementations by users.
Description is updated as per your previous comment.

**Comment by atanass** [comment]
Thank you for taking the time to look into this @aaronstillwell! I'd like to make it as clear, simple and robust as possible for users. Looking forward for any further recommendations and reviews.

**Comment by otremblay** [comment]
Hi, apologies for the long delay in response, do you think it'd be possible to use save_cache and restore_cache instead of workspaces? The reason for this is that if people are already using workspaces it could get them in a wonky state, for one, and for bigger workspaces, attaching twice can be a pretty significant operation.

The way caches can be specified by key means it could be much more robust. Thanks for the PR!

**Comment by aaronstillwell** [comment]
> Hi, apologies for the long delay in response, do you think it'd be possible to use save_cache and restore_cache instead of workspaces? The reason for this is that if people are already using workspaces it could get them in a wonky state, for one, and for bigger workspaces, attaching twice can be a pretty significant operation.
> 
> The way caches can be specified by key means it could be much more robust. Thanks for the PR!

Very good latter point... unintended consequences could also be unintentionally large credit burn - for example an large workspace being unwittingly restored to a self-hosted runner.

Using caches should be negligble for such a small amount of data in any case.

Some thinking out loud regarding moving this behaviour to a cache:
- I guess we need to have a [pipeline id](https://circleci.com/docs/pipeline-variables/#pipeline-values) in the cache key? Otherwise thread data from a previous pipeline would be loaded and a new cache would not be created
- If this is the case, each pipeline creates a new cache key, but even workflow reruns may have unexpected behaviour? I guess we could use `$CIRCLE_WORKFLOW_ID` in the path to the thread info files to prevent this?
- Every run would now create a cache entry and we usually recommend customers have a longer cache retention period that workflows. If a customer has 15 days cache retention and runs `notify` 10 times per day, that's 150 cache entries lingering around? Although given the tiny amount of data being stored, never likely to be a huge problem for anyone. Regardless, it feels pertinent that we have the option of opting-in to this functionality via an additional parameter like `persist_thread_information` and document accordingly.

Any additional objections/observations @atanass @otremblay ?

**Comment by atanass** [comment]
I changed the workspace persistence with cache storage. I ran a few corner case tests to cover complicated user inputs.
I've intentionally included the thread_id parameter value in the cache key name because it would ensure there are no inconsistency of cache content in case of parallel job invocation.

File: src/commands/notify.yml [hunk 3]
```diff
@@ -77,6 +77,13 @@ parameters:
     type: string
     default: Slack - Sending Notification
     description: Specify a custom step name for this command, if desired
+  thread_id:
+    type: string
+    default: ""
+    description: |
+      When set, the first `notify` with a given `thread_id` will appear as a regular slack message.
+      Any subsequent `notify` usage with the same identifier will be posted within the initial message's thread.
+      `thread_id` should be set to any arbitrary string to help you identify different threads. See examples for more information.
```
Comment (aaronstillwell) [inline src/commands/notify.yml:86]: @atanass could we add a quick note here to mention that enabling this parameter will use a small amount of caching?

**Comment by atanass** [comment]
> > Thanks guys! I changed the workspace persistence with cache storage. I ran a few corner case tests to cover complicated user inputs. I've intentionally included the thread_id parameter value in the cache key name because it would ensure there are no inconsistency of cache content in case of parallel job invocation.
> 
> @atanass one quick question - do you have a `save_cache` example open you can look at to check the size of the resulting cache in your example? I'm sure its tiny as to be a mute point but would be good to sanity check.

Yes, here's a snapshot:
image

File: src/commands/notify.yml (hunk 3 above)
Comment (atanass) [reply src/commands/notify.yml:86]: Sure thing.

**Comment by aaronstillwell** [comment]
> > > Thanks guys! I changed the workspace persistence with cache storage. I ran a few corner case tests to cover complicated user inputs. I've intentionally included the thread_id parameter value in the cache key name because it would ensure there are no inconsistency of cache content in case of parallel job invocation.
> > 
> > 
> > @atanass one quick question - do you have a `save_cache` example open you can look at to check the size of the resulting cache in your example? I'm sure its tiny as to be a mute point but would be good to sanity check.
> 
> Yes, here's a snapshot: image

Ok fair to say the cache usage is... negligible

**Review by otremblay — APPROVED** [review]

**Comment by orb-publisher** [comment]
Your orb has been published to the CircleCI Orb Registry.
You can view your published orb on the CircleCI Orb Registry at the following link: 
https://circleci.com/developer/orbs/orb/circleci/slack?version=4.13.1
